serveradmin/nessus/models.py

```diff
@@ -0,0 +1,278 @@
+        self.secret_key = secret_key
+        self.port = port
+        self.url = url
+        if url is None:
+            self.hostname = hostname
+            self.base = 'https://{hostname}:{port}'.format(hostname=self.hostname, port=self.port)
+        else:
+            self.base = self.url
+        self.log = log
+        self.log_level = log_level
+
+        if log_level == 'WARNING':
+            logging.basicConfig(filename=self.log, format='%(asctime)s - [ %(levelname)s ]: %(message)s', level=logging.WARNING)
+        elif log_level == "ERROR":
+            logging.basicConfig(filename=self.log, format='%(asctime)s - [ %(levelname)s ]: %(message)s', level=logging.ERROR)
+        else:
+            logging.basicConfig(filename=self.log, format='%(asctime)s - [ %(levelname)s ]: %(message)s', level=logging.INFO)
+
+        self.api_token = ''
+
+        self.session = requests.Session()
+        self.session.verify = False
+        self.session.stream = True
+        self.session.headers = {
+            'Origin': self.base,
+            'Accept-Encoding': 'gzip, deflate, br',
+            'Accept-Language': 'en-US,en;q=0.8',
+            'User-Agent': 'Nessus-Helper',
+            'Content-Type': 'application/json',
+            'Accept': 'application/json, text/javascript, */*; q=0.01',
+            'Referer': self.base,
+            'X-Requested-With': 'XMLHttpRequest',
+            'Connection': 'keep-alive',
+            'X-Api-Token': self.api_token
+        }
+
+        self.sess_token = self.login()
+
+        if len(self.sess_token) > 5:
+            self.session.headers['X-Cookie'] = "token=" + self.sess_token
+        elif self.access_key and self.secret_key:
+            logging.info('Using API keys')
+            self.api_keys = True
+            self.session.headers['X-ApiKeys'] = 'accessKey={}; secretKey={}'.format(self.access_key, self.secret_key)
+
+
+    def login(self):
+        """
+        Log user into nessus.
+
+        return string auth token that needs to be send with every later request in a cookie X-Cookie: token=<token>.
+        """
+        data = {"username": self.user, "password": self.password}
+        response = self.request('/session', method='POST', data=json.dumps(data))
+        if response.status_code == 200:
+            logging.info('Logged in to Nessus using password authentication and X-Api-Token - {}'.format(self.api_token))
+            return json.loads(response.text)['token']
+        elif "Invalid Credentials" in response.text:
+            logging.error('Invalid credentials provided! Cannot authenticate to Nessus.')
+            raise Exception('[FAIL] Invalid credentials provided! Cannot authenticate to Nessus.')
+        else:
+            logging.error('Couldn\'t authenticate! Error returned by Nessus: {}'.format(json.loads(response.text)['error']))
+            raise Exception('[FAIL] Couldn\'t authenticate! Error returned by Nessus: {}'.format(json.loads(response.text)['error']))
```

Comment:
Check for Invalid Credentials and Status Code != 200 and then continue outside of the if